src/javaharness/java/arcs/android/client/RemotePec.java

```diff
@@ -27,38 +31,76 @@ public void onMessage(String message) {
       };
 
   @Inject
-  RemotePec(ArcsServiceBridge bridge, PECInnerPortFactory pecInnerPortFactory,
+  RemotePec(
+      ArcsServiceBridge bridge,
+      PECInnerPortFactory pecInnerPortFactory,
       PortableJsonParser jsonParser) {
     this.bridge = bridge;
     this.pecInnerPortFactory = pecInnerPortFactory;
     this.jsonParser = jsonParser;
   }
 
-  public void init(String arcId, String pecId, String recipe, Particle particle) {
+  /**
+   * Starts a new arc running the given recipe. The given particle implementation is attached to
+   * that arc.
+   */
+  public void runArc(String recipe, Particle particle) {
     if (pecInnerPort != null) {
       throw new IllegalStateException("PEC has already been initialized.");
     }
 
-    pecInnerPort = pecInnerPortFactory.createPECInnerPort(pecId, /* sessionId= */ null);
-    if (particle != null) {
-      pecInnerPort.mapParticle(particle);
-    }
+    IdGenerator idGenerator = IdGenerator.newSession();
+    arcId = Id.newArcId();
+    pecId = idGenerator.newChildId(arcId, "pec");
+    Id particleId = idGenerator.newChildId(pecId, "particle");
+
+    particle.setId(particleId.toString());
+    particle.setJsonParser(jsonParser);
+
+    pecInnerPort =
+        pecInnerPortFactory.createPECInnerPort(pecId.toString(), idGenerator.getSessionId());
+    pecInnerPort.mapParticle(particle);
 
     bridge
         .connectToArcsService()
         .thenAccept(
             service -> {
               try {
                 service.startArc(
-                    arcId,
-                    pecId,
+                    arcId.toString(),
+                    pecId.toString(),
                     recipe,
-                    particle == null ? null : particle.getId(),
-                    particle == null ? null : particle.getName(),
+                    particle.getId(),
+                    particle.getName(),
                     callback);
               } catch (RemoteException e) {
                 throw new RuntimeException(e);
               }
             });
   }
+
+  /** Shuts down the running arc and remote PEC. */
+  public void shutdown() {
+    if (pecInnerPort == null) {
+      return;
+    }
+    pecInnerPort = null;
+
+    String arcId = this.arcId.toString();
+    this.arcId = null;
+
+    String pecId = this.pecId.toString();
+    this.pecId = null;
+
+    bridge
+        .connectToArcsService()
+        .thenAccept(
+            service -> {
+              try {
+                service.stopArc(arcId, pecId);
```

Comment:
maybe `this.arcId` and `this.pecId` should be set to null after the arc was actually stopped?